src/OrleansRuntime/GrainDirectory/GrainDirectoryPartition.cs

```diff
@@ -332,19 +332,22 @@ internal void RemoveGrain(GrainId grain)
         internal AddressesAndTag LookUpGrain(GrainId grain)
         {
             var result = new AddressesAndTag();
+            IGrainInfo grainInfo;
             lock (lockable)
             {
-                if (partitionData.ContainsKey(grain))
+                if (!partitionData.TryGetValue(grain, out grainInfo))
                 {
-                    result.Addresses = new List<ActivationAddress>();
-                    result.VersionTag = partitionData[grain].VersionTag;
-
-                    foreach (var route in partitionData[grain].Instances.Where(route => IsValidSilo(route.Value.SiloAddress)))
-                    {
-                        result.Addresses.Add(ActivationAddress.GetAddress(route.Value.SiloAddress, grain, route.Key, route.Value.RegistrationStatus));
-                    }
+                    return result;
                 }
             }
+
+            result.Addresses = new List<ActivationAddress>();
+            result.VersionTag = grainInfo.VersionTag;
+            foreach (var route in grainInfo.Instances.Where(route => IsValidSilo(route.Value.SiloAddress)))
```

Comment:
Then I'll just revert this commit. I think that accessing of this dictionary is single threaded in most cases, so that lock provides better performance than concurrent data structure. And cost of additional ToList will outweigh all gains from doing filtering out of lock.